qstrader/price_handler/ib_bar.py

```diff
@@ -0,0 +1,115 @@
+import os
+import datetime
+import queue
+from .base import AbstractBarPriceHandler
+
+from qstrader.service.ib import IBService
+
+from ibapi.client import EClient
+from ibapi.wrapper import EWrapper
+from ibapi.utils import iswrapper
+from ibapi.contract import *
+from ibapi.common import *
+
+
+
+#types
+from ibapi.utils import (current_fn_name, BadMessage)
+from ibapi.common import *
+from ibapi.order_condition import *
+from ibapi.contract import *
+from ibapi.order import *
+from ibapi.order_state import *
+from ibapi.execution import Execution
+from ibapi.execution import ExecutionFilter
+from ibapi.commission_report import CommissionReport
+from ibapi.scanner import ScannerSubscription
+from ibapi.ticktype import *
+
+from ibapi.account_summary_tags import *
+
+
+class IBBarPriceHandler(AbstractBarPriceHandler):
+    """
+    Designed to feed either live or historic market data bars
+    from an Interactive Brokers connection.
+
+    Uses the IBService to make requests and collect data once responses have returned.
+
+    TODO:
+        * Historic/Live mode to be set by whether QSTrader is in Backtest or Live mode
+        * IBService should be an initialization parameter
+        * Ports, etc, connection strings from config
+    """
+    def __init__(
+        self, events_queue, param_tickers, settings, mode="historic",
+        hist_end_date = datetime.datetime.now() - datetime.timedelta(days=3),
+        hist_duration="5 D", hist_barsize="1 min"
+    ):
+        self.ib_service = IBService()
+        self.ib_service.connect("127.0.0.1",4001,0)
```

Comment:
I would suggest a dynamic clientId so that you could have multiple instances running. It may be helpful to separate live trading and backtesting api connections or utilize the feature of IB that allows you to separate trades entered via the gui from the api.